File: tests/mathsplit.test.js
// task-13: node-тест splitWideMathChains (JS-зеркало mathsplit.cpp).
// Фикстуры: tests/testdata/mathsplit_cases.txt (общие с C++-тестом).
// Запуск из корня проекта: node tests/mathsplit.test.js
"use strict";

const fs = require("fs");
const path = require("path");

// Загружаем каноническую функцию (IIFE, ставит window.splitWideMathChains).
const src = fs.readFileSync(
    path.join(__dirname, "..", "frontend", "js", "mathsplit.js"),
    "utf8"
);
global.window = {};
eval(src);
if (typeof global.window.splitWideMathChains !== "function") {
    console.error("FAIL: window.splitWideMathChains не определена");
    process.exit(2);
}
const split = global.window.splitWideMathChains;

const fixture = path.join(__dirname, "testdata", "mathsplit_cases.txt");
const lines = fs.readFileSync(fixture, "utf8").split("\n").filter((l) => l.trim() !== "" && !l.startsWith("#"));

let total = 0;
let failed = 0;
lines.forEach((line, i) => {
    const no = i + 1;
    const parts = line.split("\t");
    if (parts.length !== 3 || parts[0] !== "IN") {
        console.warn(`line ${no}: malformed, skipped`);
        return;
    }
    const input = parts[1];
    const expected = parts[2];
    const actual = split(input);
    total++;
    if (actual === expected) {
        console.log(`PASS [${no}]`);
    } else {
        failed++;
        console.error(`FAIL [${no}]`);
        console.error("  input:    " + input);
        console.error("  expected: " + expected);
        console.error("  actual:   " + actual);
    }
});

if (failed === 0) {
    console.log(`ALL PASS (${total}/10)`);
    process.exit(0);
}
console.error(`FAILURES: ${failed}/${total}`);
process.exit(1);

File: tests/mathsplit_test.cpp
// task-13: unit-тест splitWideMathChains (C++).
// Фикстуры: tests/testdata/mathsplit_cases.txt (общие с node-тестом).
// Запуск из корня проекта: ./build/mathsplit_test
#include <QCoreApplication>
#include <QFile>
#include <QDir>
#include <QDebug>
#include <QStringList>

#include "mathsplit.h"

static QString findFixture()
{
    const QStringList candidates = {
        QStringLiteral("tests/testdata/mathsplit_cases.txt"),
        QStringLiteral("../tests/testdata/mathsplit_cases.txt"),
    };
    for (const QString &c : candidates) {
        if (QFile::exists(c)) {
            return QDir::toNativeSeparators(c);
        }
    }
    return QString();
}

int main(int argc, char **argv)
{
    QCoreApplication app(argc, argv);
    const QString path = findFixture();
    if (path.isEmpty()) {
        qCritical() << "FIXTURES NOT FOUND (запускай из корня проекта: ./build/mathsplit_test)";
        return 2;
    }
    QFile f{path};
    if (!f.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qCritical() << "cannot open" << path;
        return 2;
    }
    int total = 0, failed = 0;
    int lineNo = 0;
    while (!f.atEnd()) {
        const QByteArray line = f.readLine().trimmed();
        if (line.isEmpty() || line.startsWith('#')) {
            continue;
        }
        ++lineNo;
        const QStringList parts = QString::fromUtf8(line).split('\t');
        if (parts.size() != 3 || parts[0] != "IN") {
            qWarning() << "line" << lineNo << ": malformed, skipped";
            continue;
        }
        const QString input = parts[1];
        const QString expected = parts[2];
        const QString actual = splitWideMathChains(input);
        ++total;
        if (actual == expected) {
            qInfo().noquote() << QStringLiteral("PASS [%1]").arg(lineNo);
        } else {
            ++failed;
            qCritical().noquote() << QStringLiteral("FAIL [%1]").arg(lineNo);
            qCritical() << "  input:    " << input;
            qCritical() << "  expected: " << expected;
            qCritical() << "  actual:   " << actual;
        }
    }
    if (failed == 0) {
        qInfo().noquote() << QStringLiteral("ALL PASS (%1/10)").arg(total);
        return 0;
    }
    qCritical().noquote() << QStringLiteral("FAILURES: %1/%2").arg(failed).arg(total);
    return 1;
}
